statsrelay/src/backend_client/prom_client.rs

```diff
@@ -0,0 +1,522 @@
+            let client = client.clone();
+            let flush_tx = flush_tx.clone();
+            let trip = trip.clone();
+
+            tokio::spawn(async move {
+                let samples_in_batch = batch.len() as f64;
+                stats.requests_total.inc();
+                stats.samples_total.inc_by(samples_in_batch);
+
+                let write_request = construct_write_request(&stats, batch);
+                // N.B. According to the snap documentation, this compession should never fail for
+                // our use case.
+                let compressed_write_request = match compress_write_request(write_request) {
+                    Ok(c) => c,
+                    Err(e) => {
+                        stats.requests_compression_fail.inc();
+                        error!(
+                            "failed to compress prometheus WriteRequest, dropping samples: {:?}",
+                            e
+                        );
+                        return;
+                    }
+                };
+                stats
+                    .bytes_sent_total
+                    .inc_by(compressed_write_request.len() as f64);
+
+                // Attempt each request at least once
+                let mut res = client
+                    .send_write_request(compressed_write_request.clone())
+                    .await;
+                // And retry only if the tripwire has not yet been triggered.
+                while should_retry(&res) {
+                    sleep(Duration::from_millis(retry_delay_ms)).await;
+                    stats.requests_retry.inc();
+                    tokio::select! {
+                        new_res = client.send_write_request(compressed_write_request.clone()) => {
+                            res = new_res;
+                        },
+                        _ = trip.clone() => {
+                            break;
+                        },
+                    }
+                }
+
+                if let Err(e) = res {
+                    stats.requests_fail.inc();
+                    stats.samples_send_fail.inc_by(samples_in_batch);
+                    warn!(
+                        "prometheus remote write request failed, dropping samples: {:?}",
+                        e
+                    );
+                }
+
+                drop(permit);
+                drop(flush_tx);
+            });
+        }
+
+        if shutdown {
+            // Wait for all spawned tasks to terminate.
+            drop(flush_tx);
+            let _ = flush_rx.recv().await;
+            return;
+        }
+    }
+}
+
+async fn construct_batch(
+    rx: &mut mpsc::Receiver<Pdu>,
+    timeout: Duration,
+    max_batch_size: usize,
+) -> (Vec<Pdu>, bool) {
+    let mut batch = Vec::<Pdu>::with_capacity(max_batch_size);
+    let sleep = sleep(timeout);
+    tokio::pin!(sleep);
+    loop {
+        select! {
+            opdu = rx.recv() => {
```

Comment:
nit: what's `opdu` stand for? Original Protocol Data Unit? Optional?

It probably shows my background, but I tend to name this kind of thing `maybe_$name` e.g. `if let Some(pdu) = maybe_pdu { ... }`, or I'll name it the name directly and shadow it in the binding, e.g. `if let Some(pdu) = pdu { ... }`. IOW: either something that's obviously descriptive of the optionality, or just shadow it.